app/types/type-structures.ts

```diff
@@ -0,0 +1,317 @@
+    }
+
+    feed(types: Map<string, Type>) {
+        this.type = this.type.feed(types);
+        return this;
+    }
+    validate() { }
+
+    toString() {
+        return "List<" + this.type + ">";
+    }
+}
+
+// literal "class t1, t2, tn... { field:t ... }"
+export class ClassType implements Type {
+    kind = "Class"
+    constructor(readonly conformsTo: (ClassType | TypeVariable)[], readonly fields: Map<string, Type>) {
+    }
+
+    validate(): void {
+        for (const [k, t1] of this.fields.entries()) {
+            if (t1 instanceof TypeVariable) {
+                throw "ClassType.validate: free type variable";
+            }
+
+            for (const t2 of this.promisedTypes(k)) {
+                if (!t1.subtype(t2)) {
+                    throw "ClassType.validate: cannot override " + t2 + " with " + t1;
+                }
+            }
+        }
+    }
+
+    toString() {
+        return "ClassType";
+    }
+
+    promisedTypes(key: string): Type[] {
+        const result: Type[] = [];
+        for (const t of this.conformsTo) {
+            if (t instanceof TypeVariable) {
+                throw "ClassType.promisedTypes: free type variable";
+            }
+            result.push(...t.promisedTypes(key));
+        }
+        const myPromise = this.fields.get(key);
+        if (myPromise) {
+            result.push(myPromise);
+        }
+        return result;
+    }
+
+    subtype(that: Type): boolean {
+        if (that.rsubtype) {
+            return that.rsubtype(this);
+        }
+        if (this === that) {
+            return true;
+        }
+        for (const t of this.conformsTo) {
+            if (t.subtype(that)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    feed(types: Map<string, Type>) {
+        for (const [key, value] of this.fields.entries()) {
+            this.fields.set(key, value.feed(types));
+        }
+        for (let i = 0; i < this.conformsTo.length; i++) {
+            const newValue = this.conformsTo[i].feed(types);
+            if (newValue instanceof ClassType || newValue instanceof TypeVariable) {
+                this.conformsTo[i] = newValue;
+            } else {
+                throw "ClassType.feed: cannot conform to a non class type"
+            }
+        }
+        return this;
+    }
+}
+
+
+export class FunctionType implements Type {
+    kind = "Function"
+    constructor(public from: Type, public to: Type) {
+
+    }
+
+    subtype(that: Type): boolean {
+        if (that.rsubtype) {
+            return that.rsubtype(this);
+        }
+
+        if (!(that instanceof FunctionType)) {
+            return false;
+        }
+        return that.from.subtype(this.from) && this.to.subtype(that.to);
+    }
+
+    feed(types: Map<string, Type>) {
+        this.from = this.from.feed(types);
+        this.to = this.to.feed(types);
+        return this;
+    }
+    validate() { }
+}
+
+export class ThunkType implements Type {
```

Comment:
Zero argument function, used to delay execution. Will probably be used for computed fields.